### DISCUSSION TITLE
received DELETE for unknown ESP CHILD_SA

### DISCUSSION THREAD

**Discussion by beezjeh** [strongswan/strongswan#804] [Q&A]
**OS version:**
Ubuntu 20.04

**Strongswan version:**
v5.8.2

**Problem:**
We have set up a site-to-site IPsec connection, which we use to connect to SAP system and exchange data. During a data exchange, we occasionally run up to a TCP timeout. When comparing the timestamps of the TCP timeouts, it matches the error below coming from server our logs. It is not clear why we are getting a "for unknown ESP CHILD_SA" from the responder while we are the initiator.

received DELETE for unknown ESP CHILD_SA with SPI c675de1f

We also noticed that there are more rekeying attempts than we are expecting to see whenever we enable conn TWO and con THREE.
The tunnel seems to run stable whenever we cut the connection to conn TWO and conn THREE.

Amount of CREATE_CHILD_SA initiations in 3 hours with conn ONE and conn TWO enabled (2 tunnels)
cat syslog | grep "generating CREATE_CHILD_SA" | grep -v ipsec | wc -l                                                                                                                         
64

Amount of CREATE_CHILD_SA initiations in 1 day with only conn ONE enabled (1 tunnel)
cat syslog.1 | grep "generating CREATE_CHILD_SA" | grep -v ipsec | wc -l                                                                                                                       
3

**our ipsec.conf configuration**
config setup

conn SITE_ONE
     reauth = yes
     keyexchange=ikev2
     ike=aes256-sha256-modp2048
     ikelifetime=24h
     esp=aes256-sha256-modp2048
     keylife=8h
     margintime=9m
     rekeyfuzz=100%
     rekey=yes
     keyingtries=1
     authby=secret
     forceencaps=yes
     left=192.168.50.4 <our private ip>
     leftid=<our public ip>
     leftsubnet=192.168.50.0/24 <our private subnet>
     right=<customer public ip>
     rightid=<customer public ip>
     type=tunnel
     compress=no
     auto=route 
     leftauth=psk
     rightauth=psk
     leftfirewall=yes
     lifetime=8h
     keyingtries=%forever
     dpddelay=10s
     dpdtimeout=30s
     dpdaction=restart

conn ONE
     also = SITE_ONE
     rightsubnet=x.x.x.40/32 < customer private subnet 1 > 

conn TWO
     also = SITE_ONE
     rightsubnet=x.x.x.50/32 < customer private subnet 2 > 

conn THREE
     also = SITE_ONE
     rightsubnet=x.x.x.45/32 < customer private subnet 3 >

**Their IPSec configuration**
BOVPN Virtual Interface: ALM-DATA
Remote Endpoint Type: Cloud VPN or Third-Party Gateway
Restrict tunnel MTU: 1399
VPN Routes
  Route 1
    Route To: 192.168.50.0/24
    Metric: 1
Dynamic Routing
    Configured: No
    Local IP Address: 
    Remote IP Address: 
Phase 2 Settings
  Perfect Forward Secrecy: Enabled (Diffie-Hellman Group14)
  IPSec Proposals
    Proposal 1
      Name: ESP-AES256-SHA256
      Type: ESP
      Authentication: SHA2-256
      Encryption: AES (256-bit)
      Key Expiration: 8 hours
Multicast Settings
  Multicast over tunnel: Disabled
  Origination IP: 
  Group IP: 
  Helper Addresses
BOVPN Gateway Settings
  IKE Version: IKEv2
  Credential Method: Pre-shared Key
  Endpoints
    Endpoint 1
      Local Interface: External
      Local ID: <customer public ip>
      Remote IP Address: <our public ip>
      Remote ID: <our public ip>
  Phase 1 Settings
    NAT Traversal: Enabled (20 second interval)
    Dead Peer Detection: Traffic-Based (20 second timeout, 5 max retries)
    Auto Start: No
    Transforms
      Transform 1
        Authentication: SHA2-256
        Encryption: AES (256-bit)
        SA Lifetime: 24 hours
        Key Group: Diffie-Hellman Group14

** Our logs**
Dec  6 21:01:07 juju-core-0 charon: 09[ENC] parsed CREATE_CHILD_SA response 30 [ SA No KE TSi TSr ]
Dec  6 21:01:07 juju-core-0 charon: 09[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_2048/NO_EXT_SEQ
Dec  6 21:01:07 juju-core-0 charon: 09[IKE] CHILD_SA ONE{7} established with SPIs c2ede384_i bd01fd59_o and TS 192.168.50.0/24 === x.x.x.40/32
Dec  6 21:01:08 juju-core-0 charon: 12[NET] received packet: from PUBLIC_CUSTOMER_IP[4500] to 192.168.50.4[4500] (80 bytes)
Dec  6 21:01:08 juju-core-0 charon: 12[ENC] parsed INFORMATIONAL request 18 [ D ]
Dec  6 21:01:08 juju-core-0 charon: 12[IKE] received DELETE for ESP CHILD_SA with SPI cfe1e584
Dec  6 21:01:08 juju-core-0 charon: 12[IKE] closing CHILD_SA THREE{6} with SPIs c675de1f_i (3692 bytes) cfe1e584_o (8181 bytes) and TS 192.168.50.0/24 === x.x.x.45/32
Dec  6 21:01:08 juju-core-0 charon: 12[IKE] sending DELETE for ESP CHILD_SA with SPI c675de1f
Dec  6 21:01:08 juju-core-0 charon: 12[IKE] CHILD_SA closed
Dec  6 21:01:08 juju-core-0 charon: 12[ENC] generating INFORMATIONAL response 18 [ D ]
Dec  6 21:01:08 juju-core-0 charon: 12[NET] sending packet: from 192.168.50.4[4500] to PUBLIC_CUSTOMER_IP[4500] (80 bytes)
Dec  6 21:01:08 juju-core-0 charon: 11[NET] received packet: from PUBLIC_CUSTOMER_IP[4500] to 192.168.50.4[4500] (80 bytes)
Dec  6 21:01:08 juju-core-0 charon: 11[ENC] parsed INFORMATIONAL request 19 [ D ]
Dec  6 21:01:08 juju-core-0 charon: 11[IKE] received DELETE for unknown ESP CHILD_SA with SPI c675de1f
Dec  6 21:01:08 juju-core-0 charon: 11[IKE] CHILD_SA closed

Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 13[ENC] generating CREATE_CHILD_SA request 30 [ SA No KE TSi TSr ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 13[NET] sending packet: from 192.168.50.4[4500] to PUBLIC_CUSTOMER_IP[4500] (608 bytes)
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 09[NET] received packet: from PUBLIC_CUSTOMER_IP[4500] to 192.168.50.4[4500] (480 bytes)
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 09[ENC] parsed CREATE_CHILD_SA response 30 [ SA No KE TSi TSr ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 09[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_2048/NO_EXT_SEQ
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 09[IKE] CHILD_SA ONE{7} established with SPIs c2ede384_i bd01fd59_o and TS 192.168.50.0/24 === x.x.x.40/32
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[NET] received packet: from PUBLIC_CUSTOMER_IP[4500] to 192.168.50.4[4500] (80 bytes)
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[ENC] parsed INFORMATIONAL request 18 [ D ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[IKE] received DELETE for ESP CHILD_SA with SPI cfe1e584
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[IKE] closing CHILD_SA THREE{6} with SPIs c675de1f_i (3692 bytes) cfe1e584_o (8181 bytes) and TS 192.168.50.0/24 === x.x.x.45/32
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[IKE] sending DELETE for ESP CHILD_SA with SPI c675de1f
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[IKE] CHILD_SA closed
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[ENC] generating INFORMATIONAL response 18 [ D ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 12[NET] sending packet: from 192.168.50.4[4500] to PUBLIC_CUSTOMER_IP[4500] (80 bytes)
Dec  6 21:02:10 juju-core-0 charon: 11[ENC] generating INFORMATIONAL response 21 [ ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 11[NET] received packet: from PUBLIC_CUSTOMER_IP[4500] to 192.168.50.4[4500] (80 bytes)
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 11[ENC] parsed INFORMATIONAL request 19 [ D ]
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 11[IKE] received DELETE for unknown ESP CHILD_SA with SPI c675de1f
Dec  6 21:02:10 juju-core-0 ipsec[3377970]: 11[IKE] CHILD_SA closed

**Comment by tobiasbrunner** [comment]
That log is too short to say much. But the other peer definitely sends DELETEs incorrectly. To delete a CHILD_SA, only a DELETE for the inbound IPsec SA should be sent, the responder is to send back a DELETE for the corresponding other SA ([RFC 7296, section 1.4.1](https://www.rfc-editor.org/rfc/rfc7296.html#section-1.4.1), you actually see that in the `received DELETE` and `sending DELETE` log messages when strongSwan handles the first DELETE request). Your peer sends DELETEs for both and the second causes the log message you referred to because the first DELETE already caused the removal of the complete CHILD_SA.

That it sends DELETEs might be incorrect in the first place if strongSwan initiated the rekeying (we only see a rekeying for _ONE_ and DELETEs for _THREE_ in this short log excerpt, which are hopefully not related, so not sure). Only the initiator (or winner, in case of a collision) of a rekeying should send a DELETE for the old CHILD_SA. So you might want to check for that and report those issues to the vendor of the other IKEv2 implementation.

> We also noticed that there are more rekeying attempts than we are expecting to see whenever we enable conn TWO and con THREE.

Who initiates those exchanges? Are they even due to rekeyings? Again, the log is too short to say much about this. But the combination of `auto=route` with `reauth=yes` and `dpdaction=restart` is not ideal as it might cause duplicate CHILD_SAs (with each setting there might be times when no IPsec SA is installed and additional acquires might get triggered by the kernel on matching traffic, triggering additional CHILD_SAs). `reauth=no` to use proper IKE rekeying (see [ExpiryRekey](https://wiki.strongswan.org/projects/strongswan/wiki/ExpiryRekey) for details) and `dpdaction=clear` are better choices when using `auto=route`.
